daemon/containerd/cache.go

```diff
@@ -2,11 +2,86 @@ package containerd
 
 import (
 	"context"
+	"reflect"
+	"strings"
 
+	"github.com/docker/docker/api/types/container"
+	imagetype "github.com/docker/docker/api/types/image"
 	"github.com/docker/docker/builder"
+	"github.com/docker/docker/image"
 )
 
 // MakeImageCache creates a stateful image cache.
 func (i *ImageService) MakeImageCache(ctx context.Context, cacheFrom []string) (builder.ImageCache, error) {
-	panic("not implemented")
+	images := []*image.Image{}
+	for _, c := range cacheFrom {
+		im, err := i.GetImage(ctx, c, imagetype.GetImageOpts{})
+		if err != nil {
+			return nil, err
+		}
+		images = append(images, im)
+	}
+	return &imageCache{images: images, c: i}, nil
+}
+
+type imageCache struct {
+	images []*image.Image
+	c      *ImageService
+}
+
+func (ic *imageCache) GetCache(parentID string, cfg *container.Config) (imageID string, err error) {
+	ctx := context.TODO()
+	parent, err := ic.c.GetImage(ctx, parentID, imagetype.GetImageOpts{})
+	if err != nil {
+		return "", err
+	}
+
+	for _, localCachedImage := range ic.images {
+		if isMatch(localCachedImage, parent, cfg) {
+			return localCachedImage.ID().String(), nil
+		}
+	}
+
+	children, err := ic.c.Children(ctx, parent.ID())
```

Comment:
Apologies for not already knowing (or knowing exactly where to look :see_no_evil:), but is this call already returning results in "reverse created" order such that we make sure the "freshest" cache is considered first? :sweat_smile:

Reply:
Aha, nope:

https://github.com/moby/moby/blob/563fc92dd46794dd8b8f846896aceda60dc36d5c/daemon/containerd/image_children.go#L17-L82

So we probably need a similar `TODO` like this one?

https://github.com/moby/moby/blob/563fc92dd46794dd8b8f846896aceda60dc36d5c/daemon/containerd/image_list.go#L39